### PR TITLE
[ISSUE-478] XmlRecorder records all events for all neurons

### LINKED ISSUE #478: XmlRecorder records all events for all neurons instead of a single neuron
This problem corresponds to a subtask aimed at resolving Issue 240.

In the initial/first step, we focuses on migrating the variable registration functionality from the Recorder class to the appropriate variable owner classes. Currently, we enabled the XmlRecorder to record all events for a single neuron using the EventBuffer data structure. Now we want it records all events for all neurons.

In this issue, our main concern is still with the XmlRecorder.

### PR DESCRIPTION
**Pull request by xiangli6**
Closes #478 

This PR creates a variable table within the XmlRecorder, which contains a list of variables that need to be recorded. With this enhancement, the XmlRecorder becomes capable of capturing and storing all event information for all neurons.

### CODE REVIEW

File: Simulator/Recorders/XmlRecorder.h [hunk 1]
```diff
@@ -67,49 +67,59 @@ class XmlRecorder : public Recorder {
    }
 
    // Getter method for neuronName_ (only included during unit tests)
```
Comment (stiber) [inline Simulator/Recorders/XmlRecorder.h:69]: Let's use "accessor" instead of "getter".

File: Simulator/Recorders/XmlRecorder.h [hunk 2]
```diff
@@ -67,49 +67,59 @@ class XmlRecorder : public Recorder {
...
+      {
+         variableName_ = name;
+         variableLocation_ = std::shared_ptr<EventBuffer>(&location, [](EventBuffer *) {
+         });
+      }
+   };
+
+   // A list of variables
+   // the variableTable_ stores all the variables information that need to be recorded
+   std::vector<variableInfo> variableTable_;
```
Comment (stiber) [inline Simulator/Recorders/XmlRecorder.h:113]: We're including `<vector>`; don't need to use `std::`.

File: Simulator/Recorders/XmlRecorder.h [hunk 3]
```diff
@@ -67,49 +67,59 @@ class XmlRecorder : public Recorder {
...
+
+   // history of accumulated event for all neurons
+   std::vector<vector<uint64_t>> neuronsHistory_;
```
Comment (stiber) [inline Simulator/Recorders/XmlRecorder.h:116]: Should this just be renamed something like `variablesHistory`, or does it need a bigger change?

File: Simulator/Recorders/XmlRecorder.h [hunk 4]
```diff
@@ -67,49 +67,59 @@ class XmlRecorder : public Recorder {
...
 
    // a file stream for xml output
    ofstream resultOut_;
 
    string toXML(string name, vector<uint64_t> singleNeuronBuffer_) const;
```
Comment (stiber) [inline Simulator/Recorders/XmlRecorder.h:121]: Maybe `singleVariableHistory`?

File: Simulator/Recorders/XmlRecorder.cpp [hunk 5]
```diff
@@ -76,27 +76,33 @@ void XmlRecorder::term()
 /// @param[in] neurons    The entire list of neurons.
 void XmlRecorder::compileHistories(AllVertices &vertices)
 {
-   if (singleNeuronEvents_ == nullptr) {
-      return;
+   for (int iNeuron = 0; iNeuron < variableTable_.size(); iNeuron++) {
```
Comment (stiber) [inline Simulator/Recorders/XmlRecorder.cpp:79]: Change all references to the word "neuron" to "variable", I think.

File: docs/Developer/ClassDiagrams/graphitti.puml [hunk 6]
```diff
@@ -856,7 +856,8 @@ class VerticesFactory {
 class XmlRecorder {
 	+XmlRecorder()
 	+{static} Create() : Recorder*
-	#spikesHistory_ : VectorMatrix
+	#neuronsHistory_ : vector<vector<uint64_t>>
```
Comment (stiber) [inline docs/Developer/ClassDiagrams/graphitti.puml:859]: Name change.

File: docs/Developer/ClassDiagrams/graphitti.puml [hunk 7]
```diff
@@ -867,6 +868,13 @@ class XmlRecorder {
 	+printParameters() : void
 	+saveSimData(const AllVertices& vertices) : void
 	+term() : void
+	+registerVariable(string varName, EventBuffer &recordVar) : void
+
+	/' methods only used for unit test '/
+	+XmlRecorder(string fileName_)
+	+getNeuronName(int numIndex) const : string
+	+getSingleNeuronEvents(int numIndex) const : &EventBuffer
```
Comment (stiber) [inline docs/Developer/ClassDiagrams/graphitti.puml:876]: Name change.

File: docs/Developer/ClassDiagrams/recorder.puml [hunk 8]
```diff
@@ -83,24 +83,11 @@ class RecorderFactory {
...
-	+printParameters() : void
-	+saveSimData(const AllVertices& neurons) : void
-}
-
-
 class XmlRecorder {
 	+XmlRecorder()
 	+{static} Create() : Recorder*
-	#spikesHistory_ : VectorMatrix
+	#neuronsHistory_ : vector<vector<uint64_t>>
```
Comment (stiber) [inline docs/Developer/ClassDiagrams/recorder.puml:89]: Name change.

File: docs/Developer/ClassDiagrams/recorder.puml [hunk 9]
```diff
@@ -111,6 +98,19 @@ class XmlRecorder {
 	+printParameters() : void
 	+saveSimData(const AllVertices& vertices) : void
 	+term() : void
+	+registerVariable(string varName, EventBuffer &recordVar) : void
+	
+	/' methods only used for unit test '/
+	+XmlRecorder(string fileName_)
+	+getNeuronName(int numIndex) const : string
+	+getSingleNeuronEvents(int numIndex) const : &EventBuffer
```
Comment (stiber) [inline docs/Developer/ClassDiagrams/recorder.puml:106]: name change.

**Review by stiber — CHANGES_REQUESTED** [review]
Mostly minor naming changes. Hopefully, at this point, everything will generically refer to "variables".

File: Simulator/Recorders/XmlRecorder.h (hunk 1 above)
Comment (xiangli6) [reply Simulator/Recorders/XmlRecorder.h:69]: How about the getter method name? Should I also change the method name getHistory() to accessHistory()?

File: Simulator/Recorders/XmlRecorder.h (hunk 4 above)
Comment (xiangli6) [reply Simulator/Recorders/XmlRecorder.h:121]: The variable "neuronsHistory_" holds the cumulative value of all the variables in "variableTable_". However, at present, "variableTable_" only includes information about a single variable called "vertexEvent_" in the "vertices" class.  So I am uncertain whether I should use "variablesHistory_" or "singleVariableHistory". In the next step, I may create another buffer to store other types of variable such as constant variable. In term of different types of variable, So another possible name could be vertexBuffer_.  Any suggestion?

File: Simulator/Recorders/XmlRecorder.cpp (hunk 5 above)
Comment (xiangli6) [reply Simulator/Recorders/XmlRecorder.cpp:79]: This is a nested for loop to iterate the variableTable_ . Would I use rowIndex and columnIndex for each loop?

File: Simulator/Recorders/XmlRecorder.h (hunk 3 above)
Comment (xiangli6) [reply Simulator/Recorders/XmlRecorder.h:116]: “variablesHistory” sounds good.

**Comment by xiangli6** [comment]
By overloading the registerVariable() function, it becomes possible for the Xml recorder to record a vector of EventBuffer.

**Review by stiber — CHANGES_REQUESTED** [review]

**Review by stiber — APPROVED** [review]

**Pull request merged**